src/components/profile.ts

```diff
@@ -0,0 +1,170 @@
+  program: 'Your program.',
+  specialization: 'Your specialization.'
+};
+
+interface userCustomization {
+  reason: string;
+  parsedDescription?: string;
+}
+
+enum validatedFields {
+  birthdate = 'birthdate',
+  term = 'term',
+  faculty = 'faculty',
+  year = 'year'
+}
+
+export const validUserCustomization = (
+  customization: keyof typeof configMaps,
+  description: string
+): userCustomization => {
+  if (!(Object.keys(configMaps) as Array<keyof typeof configMaps>).includes(customization)) {
+    return {
+      reason: `Invalid customization selection. Must be one of **${(Object.keys(configMaps) as Array<
+        keyof typeof configMaps
+      >).map((key) => ' ' + key)}**`
+    };
+  }
+  let parsedDescription = description;
+  if (customization !== validatedFields.term) {
+    // convert first letter to capital in case the user doesnt use proper capitalization
+    parsedDescription = description.toLowerCase();
+    parsedDescription = parsedDescription[0].toUpperCase() + parsedDescription.slice(1);
+  }
+  switch (customization) {
+    case validatedFields.birthdate:
+      if (!validBirthdates.includes(parsedDescription)) {
+        return { reason: 'Invalid birthdate. Must be one of : ' + validBirthdates.join(', ') + '.' };
+      }
+      break;
+    case validatedFields.term:
+      // MASc is the only term that is not full upper case
+      if (description.toLowerCase() === 'masc') {
+        parsedDescription = 'MASc';
+      } else {
+        parsedDescription = description.toUpperCase();
+      }
+      if (!validTerms.includes(parsedDescription)) {
+        return { reason: 'Invalid term. Must be one of : ' + validTerms.join(', ') + '.' };
+      }
+      break;
+    case validatedFields.faculty:
+      if (!validFaculties.includes(parsedDescription)) {
+        return { reason: 'Invalid faculty. Must be one of : ' + validFaculties.join(', ') + '.' };
+      }
+      break;
+    case validatedFields.year:
+      if (parseInt(parsedDescription) < yearStart || parseInt(parsedDescription) > yearEnd) {
+        return { reason: 'Invalid year. Must be between: ' + yearStart + ' and ' + yearEnd + '.' };
+      }
+      break;
+  }
+  return { reason: 'valid', parsedDescription };
+};
+
+export const getUserProfileById = async (userId: string): Promise<UserProfile | undefined> => {
+  const db = await openDB();
+  return await db.get('SELECT * FROM user_profile_table WHERE user_id = ?', userId);
+};
+
+// NOTE: data will always be just one customization, but we do not know which one, meaning many of
+// the loops are just to extract an unknown key-val pair
+export const editUserProfileById = async (userId: string, data: UserProfile): Promise<void> => {
+  const db = await openDB();
+  // check if a user exists in the user_profile_table already
+  const res = await db.get(`SELECT COUNT(*) as found FROM user_profile_table where user_id = ?`, userId);
+  const user = res.found == 1;
+  let query;
+  if (user) {
```

Comment:
why not just `if(user.found==1)`

Reply:
also something something javascript something something triple-equals

Reply:
you're right, setting user to a boolean doesn't make sense. also yes triple equals oops 😆